R/bambu-extendAnnotations-utilityCombine.R

```diff
@@ -1,252 +1,302 @@
+    verbose){
+    options(scipen = 999) #maintain numeric basepair locations not sci.notfi.
+    start.ptm <- proc.time()
+    featureTibbleList <- 
+        bplapply(seq_along(readClassList), function(sample_id){
+        extractFeaturesFromReadClassSE(readClassSe = readClassList[[sample_id]],
+        sample_id = sample_id, 
+        min.readCount = min.readCount, 
+        min.readFractionByGene = min.readFractionByGene, 
+        min.geneFDR = min.geneFDR, 
+        min.txFDR = min.txFDR)}, BPPARAM = bpParameters)
+    end.ptm <- proc.time()
+    if (verbose) message("creating spliced feature tibble objects for all 
+        samples in ", round((end.ptm - start.ptm)[3] / 60, 1)," mins.")
+    ## update combinedFeatureTibble by sample, as at each step 
+    ## start and end are updated iteratively as the fead count weighted average
+    start.ptm <- proc.time()
+    listIDs <- gtools::permute(seq_along(featureTibbleList))
+    my_group_size <- 10
+    indexList <- as_tibble(listIDs) %>% 
+        group_by( (row_number()-1) %/% my_group_size) %>%
+        nest %>% pull(data)
+    combinedFeatureTibbleList <- bplapply(seq_along(indexList), function(g){
+        indexVec <- indexList[[g]]$value
+        return(sequentialCombineFeatureTibble(featureTibbleList[indexVec],
+           indexVec, intraGroup = TRUE))
+    }, BPPARAM = bpParameters)
+    combinedFeatureTibble <- 
+        sequentialCombineFeatureTibble(combinedFeatureTibbleList, 
+            indexList = NULL, intraGroup = FALSE) 
+    combinedFeatureTibble <- updateStartEndReadCount(combinedFeatureTibble)
+    end.ptm <- proc.time()
+    if (verbose) message("combing spliced feature tibble objects across all
+        samples in ", round((end.ptm - start.ptm)[3] / 60, 1)," mins.")
+    return(combinedFeatureTibble)
 }
 
 
 
+#' @noRd 
+updateStartEndReadCount <- function(combinedFeatureTibble){
+    # try first to change to long and then wide again
+    combinedFeatureTibble <- combinedFeatureTibble %>% mutate(rowID = row_number())
+    
+    startEndCountTibble <- combinedFeatureTibble %>% 
+        select(rowID, starts_with("start"),starts_with("end"),starts_with("readCount")) %>%
+        tidyr::pivot_longer(c(starts_with("start"),starts_with("end"),starts_with("readCount")),
+                 names_to = c(".value","set"),
+                 names_pattern = "(.*)\\.(.)") %>%
+        group_by(rowID) %>%
+        summarise(start = spatstat.geom::weighted.median(start, readCount),
+                  end = spatstat.geom::weighted.median(end, readCount),
+                  readCount = sum(readCount,na.rm = TRUE))
+    combinedFeatureTibble <- combinedFeatureTibble %>% 
+        dplyr::select(intronStarts, intronEnds, chr, strand, NSampleReadCount, 
+               NSampleReadProp, NSampleGeneFDR, NSampleTxFDR, rowID) %>%
+        full_join(startEndCountTibble, by = "rowID") %>%
+        select(-rowID)
+    return(combinedFeatureTibble)
+}
 
```

Comment:
```suggestion
updateStartEndReadCount <- function(combinedFeatureTibble){
    # try first to change to long and then wide again
    combinedFeatureTibble <- combinedFeatureTibble %>% mutate(rowID = row_number())
    
    startEndCountTibble <- combinedFeatureTibble %>% 
        select(rowID, starts_with("start"),starts_with("end"),starts_with("readCount")) %>%
        tidyr::pivot_longer(c(starts_with("start"),starts_with("end"),starts_with("readCount")),
                            names_to = c(".value","set"),
                            names_pattern = "(.*)\\.(.)") %>%
        group_by(rowID) %>% 
        mutate(sumReadCount = sum(readCount,na.rm = TRUE))
    
    startTibble <- select(startEndCountTibble, rowID, start, readCount, sumReadCount) %>% 
        arrange(start) %>%
        filter(cumsum(readCount)/sumReadCount>=0.5) %>% 
    filter(row_number()==1)
    endTibble <- select(startEndCountTibble, rowID, end, readCount, sumReadCount) %>% 
        arrange(end) %>% 
        filter(cumsum(readCount)/sumReadCount>=0.5) %>% 
        filter(row_number()==1)
    
    combinedFeatureTibble <- combinedFeatureTibble %>% 
        dplyr::select(intronStarts, intronEnds, chr, strand, NSampleReadCount, 
                      NSampleReadProp, NSampleGeneFDR, NSampleTxFDR, rowID) %>% full_join(select(startTibble, rowID, start), by = "rowID") %>% full_join(select(endTibble, rowID, end, readCount=sumReadCount), by = "rowID") %>%
        select(-rowID)
    return(combinedFeatureTibble)
}

```

Reply:
avoids to load external package for weighted median and is faster than the weighted.median function. It also returns an integer matching an existing coordinate. (need to be tested)